fix: return to initial state when reseting hash in Accordion & Tabs #11100

ncoden opened the pull request:
<!------------------------------------------------------------------------------
                    Please fill the following template.
            For more information, see the CONTRIBUTING.md document
------------------------------------------------------------------------------->

## Description
<!-------------------------------------------------------------------
│   Describe your changes in detail. Include any relevant information:
│   reasons, difficulties, links to web references, related issues...
└------------------------------------------------------------------->

When deep-linking is enabled and the user goes back in the history up to when no hash is set, the active panel or tab does not change and is still the one that was selected the first.

This pull request make Accordion and Tabs displaying the panel or tab that was active at the component initialization.

### Changes:
- Save up the active panel/tab at Accordion/Tabs initialization (`_initialAnchor`)
- Display the initial panel/tab when the hash change to `""`.

<!-------------------------------------------------------------------
│   Bugs and new features/improvements must be presented and
│   discussed in an issue first. Please create one if there is no issue
│   related to this pull request. You can skip this step for minor changes.
└------------------------------------------------------------------->
- Closes https://github.com/zurb/foundation-sites/issues/11100
- Replaces [#11101 - feat: add class for active tab when no hash is set in Tabs ](https://github.com/zurb/foundation-sites/issues/11101) (@jamiechong)

## Types of changes
<!-------------------------------------------------------------------
│   What types of changes does your code introduce?
│   Fill with [x] all the boxes that apply:
└------------------------------------------------------------------->
- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (anything that would change an existing functionality)
- [ ] Maintenance (refactor, code cleaning, development tools...)


## Checklist
<!-------------------------------------------------------------------
│   Please ensure that all the following points are respected.
│   Fill with [x] the boxes once the rule is respected.
└------------------------------------------------------------------->
- [x] I have read and follow the CONTRIBUTING.md document.
- [x] The pull request title and template are correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- ~~I have updated the documentation accordingly to my changes (if relevant).~~
- ~~I have added tests to cover my changes (if relevant).~~


<!------------------------------------------------------------------------------
            For more information, see the CONTRIBUTING.md document         
              Thank you for your pull request and happy coding ;)          
------------------------------------------------------------------------------->

ncoden committed:
fix: return to initial state when reseting hash in Accordion & Tabs #11100

When deep-linking is enabled and the user goes back in the history up to when no hash is set, display the initially-selected content/tab instead of staying with the content/tab that was selected the first.

Closes https://github.com/zurb/foundation-sites/issues/11100
Replaces https://github.com/zurb/foundation-sites/issues/11101

Co-authored-by: Jamie Chong <jamie@hippocurious.com>

SassNinja reviewed (changes requested):
@ncoden I've tested the accordion and it works – but only if an item is initially active!

If no item has initially the class `is-active` is fails.
I think we should cover that case, too.

SassNinja commented on js/foundation.tabs.js:
```diff
@@ -98,8 +104,14 @@ class Tabs extends Plugin {
+      if (anchor.length) {
         var anchorNoHash = (anchor.indexOf('#') >= 0 ? anchor.slice(1) : anchor);
         var $link = this.$element.find(`[href$="${anchor}"],[data-tabs-target="${anchorNoHash}"]`).first();
         if ($link.length) {
```
In the course of testing the tabs I noticed the focus style remains when using history back.

![tabs_deep_linking_480px_low](https://user-images.githubusercontent.com/19912475/44781330-5e3eb200-ab84-11e8-8390-34dd1439f7c0.gif)

What do you think about auto focusing the link to fix this?

```javascript
if ($link.length) {
  $link.focus();
  this.selectTab($(anchor), true);
```

ncoden replied:
This is an other issue, not related to this pull request. Why not, but I wonder what's the impact on accessibility to move the focus like this.

ncoden committed:
fix: fix initial state for Accordion without any panel active

When going back in history, the Accordion without any panel active is still considered as in its "initial state".

ncoden committed:
refactor: split DOM operations from logics in Accordion

Split DOM operations from logics in Accordion "up" and "down" methods

ncoden committed:
refactor: move "init only" Accordion logic to its own methods

ncoden committed:
docs: remove outdated comments in Accordion related to the initial state

ncoden commented:
@SassNinja I refactored the component to clean up a bit the logic and split the options processing/component logic from simple DOM manipulations. It should resolve the first bug you pointed out.

SassNinja commented:
> Could you review this again please ? :)

It's working better now: history back adjust the adressbar as expected.

But the active state is not working as expected:
when I start without an initially opened tab, click on a tab and go back in history, I don't get the original active state (none tab opened). Instead the previously clicked tab is active (although the page url has changed).

Shall I provide an example repo or is it clear what I mean?

ncoden commented:
> Shall I provide an example repo or is it clear what I mean?

I understood the issue, and through I resolved this issue. If you look at the code, you can see that `_closeAll()` is called if we go back to a state where no link is active.

ncoden commented:
Actually, I fixed this issue for the Accordion only. :p

SassNinja commented:
Just to be clear: the accordion component works as expected.
The issues described above only refers to the tabs components.

Here's a quick gist with the content of my `\test\visual\tabs\deep-linking.html`
https://gist.github.com/SassNinja/c194da0830de81a510061c99ffcb8938

ncoden commented:
> Just to be clear: the accordion component works as expected.
> The issues described above only refers to the tabs components.

Yeah I got it. I need to do the same refactor in Tabs as I did in Accordion to resolve it. I'll take care of this sunday.

ncoden committed:
feat: add "Tabs._collapse()" to collapse the active tab

ncoden committed:
refactor: clean up the Tab initialization

ncoden committed:
fix: close the active Tab when going back to a "tab-less" history state

See https://github.com/zurb/foundation-sites/pull/11477#issuecomment-419678146

ncoden committed:
docs: fix "zplugin" typo in Accordion & Tabs events

ncoden commented:
@SassNinja Done.

ncoden commented:
@SassNinja 👋 Could you test it again?

SassNinja commented:
@ncoden lgtm 👍 

Just two things don't look nice
- focus state doesn't get reseted on history back (as mentioned earlier)
- the hash tag in the browser url doesn't get removed when I toggle a tab, so that no panel is opened (when using `data-active-collapse="true"`)

But nothing related to this PR – so good to merge imo.
